### PR TITLE
feat: migrate to autocomplete.js v1 [BREAKING CHANGES]

### LINKED ISSUE #76: Migrate to autocomplete.js v1
Currently we use the outdated autocomplete.js v0.
The decision was made for lack of better alternative and because we knew it correctly.
As the release of the refacto is coming soon and Docsearch using it, we could think about also migrating to v1.

### PR DESCRIPTION
**Pull request by bodinsamuel**
Closes #76

## Changes
- Complete migration to autocomplete v1
- New option for theming
- `algoliasearchNetlify()` will throw if missing mandatory option
- the UI has a minimum of `400px` width

Screenshot 2020-11-17 at 14 29 09

## How to migrate
- autocomplete.js now creates its own input, so we no longer support input selector. 
  solution: remove your input, create an empty node, e.g:`<div id="#search"></div>` and use `selector: "#search"`
- `color` property has been removed in favor of `theme` for improved theming ability.
- `silenceWarnings` has been removed
- `autocomplete` property has been removed in favor of `selector` and `hitsPerPage` at the top level
- ui no longer adapt to size of the input and follow autocomplete.js `panelPlacement: 'input-wrapper-width'`, to improve this behavior you can adapt the CSS on your own with media queries.

### CODE REVIEW

File: frontend/src/AutocompleteWrapper.ts [hunk 1]
```diff
@@ -1,206 +1,144 @@
...
-          `[Algolia] No input matched our default selector ${inputSelectorText}`,
-          'The integration needs a search input to be active on your page. You can either:',
-          '- add an input that matches the selector',
-          '- or modify `autocomplete.inputSelector` to match your search input.',
-        ].join('\n')
-      );
+  render() {
+    const $input = document.querySelector(this.options.selector) as HTMLElement;
+    if (!$input) {
+      console.error('[algoliasearch netlify] no inputs found');
```
Comment (sbellone) [inline frontend/src/AutocompleteWrapper.ts:36]:
```suggestion
      console.error(
        `[algoliasearch-netlify] no element ${this.options.selector} found`
      );
```

File: frontend/src/AlgoliasearchNetlify.ts [hunk 2]
```diff
@@ -1,66 +1,49 @@
...
+
+function algoliasearchNetlify(_options: Options) {
+  const options = {
+    ...defaultOptions,
+    ..._options,
+  };
+  for (const key of mandatory) {
+    if (options[key]) continue;
+
+    throw new Error(`Missing mandatory key: ${key}`);
```
Comment (sbellone) [inline frontend/src/AlgoliasearchNetlify.ts:32]:
```suggestion
    throw new Error(`[algoliasearch-netlify] Missing mandatory key: ${key}`);
```

File: frontend/src/AutocompleteWrapper.ts [hunk 3]
```diff
@@ -1,206 +1,144 @@
...
-      );
-
-      const handleSelected = this.createHandleSelected();
-      aa.on('autocomplete:selected', handleSelected);
-
-      return aa;
+    const instance = autocomplete<AlgoliaRecord>({
+      container: $input,
+      autoFocus: false,
+      placeholder: 'Search...',
```
Comment (sbellone) [inline frontend/src/AutocompleteWrapper.ts:43]: Would be cool to make it configurable

File: frontend/src/AutocompleteWrapper.ts [hunk 4]
```diff
@@ -1,206 +1,144 @@
...
+      debug: this.options.debug,
+      openOnFocus: true,
```
Comment (sbellone) [inline frontend/src/AutocompleteWrapper.ts:45]: This one too please, I personally don't like to see random results before I type anything

File: frontend/src/AutocompleteWrapper.ts [hunk 5]
```diff
@@ -1,206 +1,144 @@
...
-  }
-
-  private computeNbSnippetWords(inputWidth: number): number {
-    if (inputWidth < XS_WIDTH) return 0;
-    if (inputWidth < SM_WIDTH) return 3 + Math.floor(inputWidth / 35);
-    return Math.floor(inputWidth / 20);
-  }
-
-  private getDropdownTemplates(poweredBy: boolean): { footer?: string } {
-    if (!poweredBy) return {};
```
Comment (sbellone) [inline frontend/src/AutocompleteWrapper.ts:154]: This breaks the `poweredBy` option which is not taken into account anymore

File: frontend/src/types/record.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+export type AlgoliaRecord = {
+  objectID: string;
+
+  url: string;
+  origin: string;
+  title: string;
```
Comment (sbellone) [inline frontend/src/types/record.ts:6]:
```suggestion
  title?: string;
```

File: frontend/src/types/record.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  hierarchy?: string[];
+  hierarchyObj?: {
+    [lvl: string]: string;
+  };
+  currentHierarchy?: string;
+
+  contentLength?: number;
+
+  recordUrl?: string;
```
Comment (sbellone) [inline frontend/src/types/record.ts:29]:
```suggestion
```

**Review by sbellone — COMMENTED** [review]
Looks promising!
Main issue is that the results are not closed when we click away and stay open forever.

File: frontend/src/AutocompleteWrapper.ts (hunk 4 above)
Comment (francoischalifour) [reply frontend/src/AutocompleteWrapper.ts:45]: As a recommendation, `openOnFocus` works well when you want to display recent searches, hot searches or a site map. I wouldn't use it by default in this integration.

**Comment by francoischalifour** [comment]
> Main issue is that the results are not closed when we click away and stay open forever.

@sbellone That's because the `debug` option is used in development in your integration. I wouldn't recommend using it as a default though, as it's mostly useful to inspect the results panel when templating your items. It shouldn't be used in production—in fact, you should see a warning.

We talk about this in the "[Debugging](https://algolia-autocomplete.netlify.app/docs/debugging)" section of our docs.

File: frontend/src/AutocompleteWrapper.ts (hunk 3 above)
Comment (bodinsamuel) [reply frontend/src/AutocompleteWrapper.ts:43]: indeed

File: frontend/src/AutocompleteWrapper.ts (hunk 5 above)
Comment (bodinsamuel) [reply frontend/src/AutocompleteWrapper.ts:154]: good catch

File: frontend/src/types/record.ts (hunk 6 above)
Comment (bodinsamuel) [reply frontend/src/types/record.ts:6]: title can not be empty (even if it was set like that in the original types) the code will always output at least `""` (empty string)

Comment (sbellone) [reply frontend/src/types/record.ts:6]: Which "code"?

Comment (bodinsamuel) [reply frontend/src/types/record.ts:6]: cleanWhitespaces()

**Comment by bodinsamuel** [comment]
- Disabled openOnFocus and added an option
- Added option for placeholder
- Fixed poweredBy
- Fixed types

**Review by sbellone — APPROVED** [review]
🚀

**Comment by sbellone** [comment]
I've just found out that it doesn't work correctly if the index has no `attributesToSnippet` (only the title is displayed in that case).
The current version of the UI is requesting them at runtime, I think this new version should do the same.

**Comment by bodinsamuel** [comment]
> if the index has no attributesToSnippet 

All Netlify indices are pre configured with `attributesToSnippet` by default

**Comment by sbellone** [comment]
> All Netlify indices are pre configured with attributesToSnippet by default

Automatically? That's not what I have observed. They currently have `content` as `attributesToSnippet` because we set it in the initial index settings, but that's it.
Here your UI expect `description` to be snippeted and it's not the case by default so the description will never be displayed until the users put it as an attribute to snippet. It fallbacks on the `content`.

What's the problem of sending them at runtime just in case since we rely on them for the display?

**Comment by bodinsamuel** [comment]
> Here your UI expect description

ah, we should just change our default index settings then.

**Comment by sbellone** [comment]
Ok but we can also have the UI fallback to the actual fields if no snippetting available, to not have the UI broken because of index settings

**Comment by bodinsamuel** [comment]
> to not have the UI broken because of index settings

Did the change but index settings are very tied to the UI, they change it they assume it imo.

**Comment by sbellone** [comment]
True, so a fallback on raw fields if no snippets available permits to let user controlling the length of the snippets actually. As you prefer!

**Pull request merged**